pkg/distributor/forwarding/forwarding.go

```diff
@@ -253,7 +262,25 @@ type tsByTargets map[string]tsWithSampleCount
 
 // copyToTarget copies the given time series into the given target and does the necessary accounting.
 // The time series is deep-copied, so the passed in time series can be returned to the pool without affecting the copy.
-func (t tsByTargets) copyToTarget(target string, ts mimirpb.PreallocTimeseries, pool *pools) {
+func (t tsByTargets) copyToTarget(target string, dontForwardBeforeTs int64, ts mimirpb.PreallocTimeseries, pool *pools) (err error) {
+	if dontForwardBeforeTs > 0 && samplesNeedFiltering(ts.TimeSeries.Samples, dontForwardBeforeTs) {
+		samplesUnfiltered := ts.TimeSeries.Samples
+		defer func() {
+			ts.TimeSeries.Samples = samplesUnfiltered
+		}()
+
+		samplesFiltered := filterSamplesBefore(samplesUnfiltered, dontForwardBeforeTs)
+		if len(samplesFiltered) < len(samplesUnfiltered) {
+			err = errSamplesTooOld
```

Comment:
In addition to returning error, we should also include such samples in `cortex_discarded_samples_total` metric.

Reply:
I think updating `cortex_discarded_samples_total` would make sense if we'd not ingest them when `ingest` is `true`, but we do. Maybe there should be a new metric to count the samples that were not forwarded due to the age?

Reply:
I think we should add new reason for discarded samples that were not ingested. Let's do that in next PR.

Reply:
https://github.com/grafana/mimir/pull/3113